### ISSUE TITLE
Accept relative paths for some file/dir config options

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#622]
e.g. for `user-stylesheet`, a path relative to `~/.config/qutebrowser` would make sense.

**Comment by lamarpavel** [comment]
So, I started to work on this and got this feature working for at least setting a user-stylesheet, with all tox-tests succeeding. As this is my first time doing anything with python I have no idea how acceptable my approach is and would be pleased if you could take a look [at what I wrote](https://github.com/lamarpavel/qutebrowser/commit/58f031630c6b04c0178f268bf5ad3803455d0c31).

This is not a general solution, I think, just a first draft for that one specific file. Once I have your approval I will try to apply this to other config files as well. And don't worry about coding style, I will read your "how to contribute" docs before I create a pull request.

**Comment by The-Compiler** [comment]
Awesome, thank you!

Some background first:

Note every `ConfigType` has a `validate` and a `transform` method.

`validate` is called to validate the config, and the value is considered valid if it doesn't raise the `configexc.ValidationError` exception. `transform` is called to get the actual value.

You added your code to `validate` which only changes the validation so that it passes - but `transform` still does the same thing. I could be mistaken (as I haven't tested it), but I think this only works "by accident" when you happen to be in `~/.config/qutebrowser` when testing (because it's a relative path).

Instead, you should use `os.path.join` if it isn't absolute in `transform` - and then maybe just call `transform` inside `validate` *before* validing the file is actually there, as other types like `FormatString` and `List` do.

If something wasn't clear, just ask! :smile:

**Comment by lamarpavel** [comment]
Alright, I have one question for now:

In the definiton of ```transform``` in class __BaseType__, which I think __File__ and __UserStyleSheet__ are derived from, it is stated that _This method can assume the value is indeed a valid value_, which I thought was a guarantee to callers that I had to respect. Wouldn't calling ```transform``` inside of ```validate``` be a violation of that guarantee?

**Comment by The-Compiler** [comment]
It's a bit more subtle then that: The callers usually don't care, other than that `transform` should never raise an exception with any string given (it's called with the strings coming from the config).

The guarantee described there is actually one the *caller* needs to follow: It should never call `transform` without using `validate` first. In other words, `transform` *can* rely on the fact that it will never get invalid values.

Here, we just decide to not use that convenience - we call `transform` in `validate`, so `transform` needs to be able to take invalid values. However here that shouldn't be an issue, as any string should be allowed for the second path of `os.path.join` as well as for `os.path.expanduser` - both don't care whether the paths actually exist.

**Comment by lamarpavel** [comment]
Okay, got it.

A general question related to python: In some classes ```transform``` is defined before ```validate``` and in some ```validate``` is defined first.
In ```UserStyleSheet``` ```validate``` is defined first. If I call ```transform``` from within ```validate``` there, which ```transform``` will be called? That of the same class, or that of the base class ```File```?

**Comment by The-Compiler** [comment]
Oh, good point! I probably should clean this up after you're done with this file (so we don't get merge conflicts).

The order doesn't matter - if a method is overridden, the overridden one is always called (and it can optionally call the base class one via `super()` - see `BoolAsk` for a good example.

**Comment by lamarpavel** [comment]
I think I got it working now, was held up by failing tox tests and lack up python knowledge. [Take a look if you can find the time](https://github.com/The-Compiler/qutebrowser/compare/master...lamarpavel:relapaths).

I initially tried calling ```transform``` within ```validate```, as you suggested, but ```transform``` is not necessarily returning strings, as expected by ```validate``` and I wasn't sure how to properly  convert types in Python.

**Comment by The-Compiler** [comment]
Sorry for the delay, I was a bit busy over the last few days!

Apart of a few small things (e.g. not checking whether `standarddir.config()` could be `None` in `validate`, it looks good to me.

I'd say let's go one step further before merging though if you agree :wink:

I think it'd be best to move all that logic to `File` so it works everywhere and the code isn't duplicated, and then do something like this instead for `UserStylesheet`:

```python
class UserStyleSheet(File):

    # [...]

    def transform(self, value):
        value = super().transform(value)
        if os.path.exists(value):
            return QUrl.fromLocalFile(path)
        else:
            data = base64.b64encode(value.encode('utf-8')).decode('ascii')
            return QUrl("data:text/css;charset=utf-8;base64,{}".format(data))

    def validate(self, value):
        if not value:
            if self._none_ok:
                return
            else:
                raise configexc.ValidationError(value, "may not be empty!")
        try:
            super().validate(value)
        except configexv.ValidationError:
            # probably a CSS, so we don't handle it as filename.
            # FIXME We just try if it is encodable, maybe we should
            # validate CSS?
            # https://github.com/The-Compiler/qutebrowser/issues/115
            try:
                value.encode('utf-8')
            except UnicodeEncodeError as e:
                raise configexc.ValidationError(value, str(e))
            return
```

If that makes sense, I'd say let's open a pull request after that change, so I can more easily comment on individual lines if there's anything left.

**Comment by lamarpavel** [comment]
As you can see in the pull request my solution for ```validate``` is more verbose than yours. I tried to get rid of the extra try-except-block but got stuck for an hour on the resulting tox failures and figured you could point me in the right direction on how to solve this more elegantly. I haven't made much use of exceptions since my Java-days at university and python is still pretty foreign to me, sorry for the inconvenience.

**Comment by lamarpavel** [comment]
I have just merged my local branch with upstream to verify all tox tests still succeed and everything works well. In qutebrowser I have tested this specifying a new user-stylesheet as

- absolute path
- relative path
- path with ~/ substitute
- default css-string
- path to nonexistant file

All of these are now working as expected (the non-existant file results in the standard qt theme without additional css applied). Is there something else I should test for? Unless I've missed something I'd say the [PR](https://github.com/The-Compiler/qutebrowser/pull/686) is ready (:

**Comment by The-Compiler** [comment]
Cool, thanks! I was a bit busy with all the other recent PRs, but I hope I'll find to test and review this again today or tomorrow.

**Comment by lamarpavel** [comment]
No sweat, I just wasn't sure if I missed something. Take your time.

**Comment by The-Compiler** [comment]
Closing this because #686 was merged :smile:
